### DISCUSSION TITLE
Controlling compressor power of Bosch Compress Heatpumps

### DISCUSSION THREAD

**Discussion by Vippis2000** [emsesp/EMS-ESP32#1678] [Ideas]
Hello everybody,
I'm new to this topic. I came to know to this project from IOBroker community. Bosch sells a solution called "Energy Manager" which has the ability to modulate the compressor's power in relation to PV production. This seems to work via KM200 LAN adapter or EMS I don't know. So has any of the developers here done this successfully? 

If not I would be give you my system for testing. My aim is to build a PV (summer time) and or dynamic electricity price based (if there is not enough PV production or winter time) strategy to control the heatpump incl. compressor to minimize grid consumption in PV oriented mode.

**Comment by proddy** [comment]
We'd love to understand more about this 'energy manager'. What you're suggesting is one of the next big features we are looking at adding to EMS-ESP. Let's collaborate.  @bbqkees

**Reply by Vippis2000** [reply]
You can find information here (German):

https://www.bosch-homecomfort.com/de/de/wohngebaeude/wissen/heizungssteuerung/der-energiemanager/

**Reply by Vippis2000** [reply]
The Bosch solution is really expensive, because it needs extra measurements for inverter/battery and grid side, unless you have one of the supported inverter models. In addition you need the Bosch Smart Home controller and a Software Upgrade Key. Total cost is around 1200€ plus installation cost.

I'm really curious how it communicates with the heat pump as I have asked Bosch. They said you only need to setup via HomeCome Easy App, which means you dont need to connect the heatpump via cable.

Can you tell me how one can reverse-engineer their solution?

**Reply by proddy** [reply]
I don't own a heat pump or even have solar panels, so we need @bbqkees's ideas on this and then we when a design and some functional specs we can start coding it

**Reply by Vippis2000** [reply]
How do we find out what entity is the one to control the compressor ? You can teach me and I can start investigating...

**Reply by proddy** [reply]
via trial and error I'm afraid. Make changes and monitor the logs. https://emsesp.github.io/docs/Troubleshooting/#im-missing-certain-data-from-an-ems-device

**Reply by Tomk211** [reply]
Is there any update on this issue? Would be interested and can support with Bosch heat pump and Pv.

**Reply by stegeb1** [reply]
I think what is missing right now is enough EMS-Bus traffic over a longer time with well know conditions, so we can start creating and testing some possible ideas on the communication between the bosch energy manager and the heat pump. The energy manager traffic originates from the modem, so easy to filter.

**Reply by nicoleise** [reply]
> How do we find out what entity is the one to control the compressor ? You can teach me and I can start investigating...

Which heat pump do you have? Of fx 7001 and 7800 there are four inputs that act as scenes. You set up in the installer menu what they should do. One is intended as a PV input which tells the hest pump to increase production because cheap energy is available.

**Comment by Vippis2000** [comment]
For this i need to have the Bosch Energy Manager Running to monitor its traffic on the EMS. But I don't have it, maybe someone else has it up and running ?

**Reply by Vippis2000** [reply]
Can you please explain to me how to enter values in these entities:

![image](https://github.com/emsesp/EMS-ESP32/assets/128602940/25ddaa2f-c928-41d5-86e9-ad1a5d41c704)

![image](https://github.com/emsesp/EMS-ESP32/assets/128602940/9afeb141-9f50-4ae6-b289-f66f1f001aa3)

One of those is labeled as "comp"

**Reply by proddy** [reply]
It is not easy, I agree. In the next version we'll add a feature to the WebUI to make this easier. There may be some additional help in https://github.com/emsesp/EMS-ESP32/issues/600

**Reply by MichaelDvP** [reply]
There were long discussions. The actual input is as described in the helptext. For inverting only you can send 0/1 and the other options stay. If you like to set a option like PV=1, you have to send a string up to PV value, not changing the others are set as `x` e.g . for PV:
`xxxxxxxxx1` or `1000000001` for INV and PV on, all others off. for setting EVU1 and keep invert unchanged it is `x1`

**Reply by Vippis2000** [reply]
Do you think there might be a undiscovered combination of values for enabling compressor power modulation?

**Comment by RaHehl** [comment]
[PV-Start.txt](https://github.com/emsesp/EMS-ESP32/files/15175610/PV-Start.txt)
That's what happens if the Compressor is started by the energymanager

Water Heating mode started at around 10:59:38

**Reply by Vippis2000** [reply]
Thanks for sharing! Was the compressor power also modulated?

**Reply by RaHehl** [reply]
In theory yes, but i think in the current szenario there was no power limit 
![image](https://github.com/emsesp/EMS-ESP32/assets/7577984/585a8c6c-415a-4cbc-b70e-b9fa2f230fe8)

![image](https://github.com/emsesp/EMS-ESP32/assets/7577984/fb5a0cc4-0bdc-46f9-bcde-66613d041ec9)

![image](https://github.com/emsesp/EMS-ESP32/assets/7577984/3e16db93-77bd-46a5-8514-6c9a59405ed1)

![image](https://github.com/emsesp/EMS-ESP32/assets/7577984/51129acc-61a3-497b-b763-55bf765bc0d6)

![image](https://github.com/emsesp/EMS-ESP32/assets/7577984/de99b385-66d8-44e3-9af8-4c3f09228a49)

![image](https://github.com/emsesp/EMS-ESP32/assets/7577984/ec7f2c78-e346-4d97-853e-4dbe2279ade1)

![image](https://github.com/emsesp/EMS-ESP32/assets/7577984/19aa7e15-2b4a-459b-a73a-ef2b0bfcd2e9)

**Reply by Vippis2000** [reply]
Does the energy manager app give any clue if there was a controlling event or something?

**Reply by Vippis2000** [reply]
So please share again a Log file on a "cloudy" day if the system has to intervent ...

**Reply by RaHehl** [reply]
"Does the energy manager app give any clue if there was a controlling event or something?"

Nope it's a total Blackbox

**Reply by popopirat** [reply]
I am also interested in the reverse engineering! Almost all of my devices are not compatible with the Bosch Energy Manager App. I have already built something similar for an electric hot water heater based on excess pv energy and now i want to do the same for the heatpump. @Vippis2000 I am still struggling to find a way to set the maximum compressor power, since the method from [this thread](https://github.com/emsesp/EMS-ESP32/issues/1717) does not work for me or I get a reading which is completely off (65 kW).

**Comment by RaHehl** [comment]
see also https://github.com/emsesp/EMS-ESP32/issues/1717

**Reply by stegeb1** [reply]
Have you figured out more on the interaction between the Bosch EnergieManager and the Heatpump. I have locked through the logs and focused on communication between gateway and boiler. There is some interesting communication, but nothing conclusive. It would simply be good to have ems bus traffic for a longer period of time to figure out some pattern. Ideally with EM off and with EM on.

**Comment by stegeb1** [comment]
My "low cost" work around until understanding how to modulate the heat pump power consumption (CSi6800AW) is to switch to "silent mode" ON (which is implemented) and then set the "silent mode power limit" of the boiler 0x08 using 0x0484 with offset 64/0x40 (UINT). {3: 30% reduction of max electric power / 70% max electric power, 4: 40% / 60%, 5: 50% / 50%, 6: 60% / 40%). This is not perfect, but can help to keep some control on the power consumption .

**Reply by Vippis2000** [reply]
> My "low cost" work around until understanding how to modulate the heat pump power consumption (CSi6800AW) is to switch to "silent mode" ON (which is implemented) and then set the "silent mode power limit" of the boiler 0x08 using 0x0484 with offset 64/0x40 (UINT). {3: 30% reduction of max electric power / 70% max electric power, 4: 40% / 60%, 5: 50% / 50%, 6: 60% / 40%). This is not perfect, but can help to keep some control on the power consumption .

I have tried the same.

Using  0x0484 with offset 64/0x40 (UINT) on my CS7000iAW does not work.

What I have found out is that in Silent Mode using the implemented hpmaxpower entity I can control compressor power for heating mode. For warmwater the HP ignores the hpmaxpower setting in silent mode.

Does your CS6800 commits to Max Power for heating and warmwater mode?

**Reply by stegeb1** [reply]
Interesting to hear that your CS7000 does no work. Do you get any response when reading the registers. I works absolutely smooth for me. Just implemented it into an homeassistant automation. (Again this is a workaround, and I am more interested in setting the max compressor power in a more controlled way). "hpmaxpower" as you suggest does not show up in emsesp for. In emsesp hpmaxpower only shows up in the device entitiy list of the CSx800 in the customization list, but not in the device entity list.  Not sure why. Do you know the EMS address, offset and datatype of hpmaxpower

**Reply by Vippis2000** [reply]
Maybe @proddy can tell you this?

**Comment by T9C5F** [comment]
A new `boiler_pv_compressor_max_power` entity was added in release 3.7.0 by commit https://github.com/emsesp/EMS-ESP32/commit/ffdcbac1e084c3e400898cb9fefdab0404fac235 - I cannot set this value. Is there any documentation or information on this?

**Reply by proddy** [reply]
why can't you set the value? Does it error? Its a unsigned integer 0-254 indicating the Kw.

**Reply by MichaelDvP** [reply]
Check in customization if the entity is listed without value. Then your ems system does not publish it. Also check with `read 8 484` the telegram.

**Reply by stegeb1** [reply]
I encounter this same issue with hpmaxpower -  can see on customization list without value - and not contained in the device entity list

**Reply by proddy** [reply]
think we need to write an article on in the doc/troubleshooting on this. Something like mentioning that if EMS-ESP doesn't pick up a valid value from a device entity (while it does show as supported under Customizations) then it's not implemented for that model. The user would need to find the telegram by looking it up on https://emsesp.org/All-Devices/#all-devices-and-entities, calling the `read` command from the System Log and posting the output. I would need to add the telegramID to the XLS first.

**Reply by stegeb1** [reply]
will really be helpful. The url is just what I was looking for. Is there any kind of an inventory, which data fields within a telegram ID have been successfully been confirmed.

**Reply by proddy** [reply]
>  Is there any kind of an inventory, which data fields within a telegram ID have been successfully been confirmed

No, we would need the community to keep that alive. You can make a start, rally some people and maintain it? I just don't have the time or energy any more

**Reply by T9C5F** [reply]
I can't set the value as it always stays at 0. It is shown in Customizations. Calling the read command also confirms this. I was just curious if I could limit the compressor power for PV and found this new entity. I didn't even know if the entity was based on the Bosch energy manager or part of the silent mode workaround. I do not have silent mode turned on. And I am not sure why the entity is called PV if it is "just" a power limit in kW in silent mode? Maybe I'm missing something, so I asked if there was any information on this. Thanks for your time!

> ems-esp:$ read 8 484
004+10:01:17.319 N 1: [emsesp] boiler(0x08) -W-> me(0x0B), HPSilentMode(0x0484), data: 01 00 23 2D 04 50 00 00 00 02 01 F6 26 1E 02 02 05 19 0A 0A 03 0D 07 00 0A
004+10:01:17.364 N 2: [emsesp] boiler(0x08) -W-> me(0x0B), HPSilentMode(0x0484), data: 01 90 00 F6 00 00 64 00 00 E1 00 00 00 00 0A 64 00 64 54 20 00 01 00 00 23 (offset 25)
004+10:01:17.406 N 3: [emsesp] boiler(0x08) -W-> me(0x0B), HPSilentMode(0x0484), data: 58 18 **00** 00 00 00 00 (offset 52)
004+10:01:17.436 N 4: [emsesp] boiler(0x08) -W-> me(0x0B), HPSilentMode(0x0484), data: <empty> (offset 59)

**Comment by nicoleise** [comment]
I think you might be going at this in a slightly weird way. Why modulate the compressor power?

What Bosch themselves are (seems to be) doing it changing the flow temperature setpoints which makes a lot more sense. 

In the installer menu, I see options to limit the offset/influence and these are by default on my device +/- 2°K. I would argue -0…+4 would be better. 

So essentially, it simply converts the building into s thermal battery, but requires either no thermostats/RTRs or some that collaborate to also increase room setpoints to actually take advantage. 

I haven't determined whether they also do something to DHW yet, but I would imagine they do. It will likely require the difference between current DHW internal temp and extra hot DHW end temp to be greater than X, and would probably be as simple as enabling the "Extra Hot Water" feature. However, they can't "just" do that as they don't require a central domestic water mixing valve, which means temps shouldn't exceed 65 °C due to the risk of scolding. So perhaps they do something "milder" like switch profile to comfort if not already selected or set more moderate DHW setpoints temporarily.

**Reply by nicoleise** [reply]
Ah - a PS: You can change the setpoint yourself through EMS, even if your boiler/HP doesn't support external setpoints (mine doesn't). 

Simply set up outdoor compensation (heat curve) but make a flat "curve" at whatever temperature you consider the maximum you want. For me it's 60 °C. So regardless of OAT, the heat pump will/would produce 60 °C. 

Now you can use the parameter "maximum flow temp" to control the flow temperature setpoint as desired. So mine currently is 48, and the heat pump respects this and outputs just around 48°. 

Just be mindful to delay/ramp changes, either in both directions (typically needed for gas boilers) and as a minimum when reducing (to avoid flow temp too hot errors).
